thirdeye/thirdeye-frontend/app/pods/components/heatmap-chart/component.js

```diff
@@ -0,0 +1,184 @@
+        let dimVals = dimValues[n] || new Set();
+
+        // conditional rollup
+        if (rollup > 0 && Object.keys(curr).length >= rollup) {
+          const topk = new Set(this._makeTopK(curr, rollup - 1));
+          curr = this._makeRollup(curr, topk, 'OTHER');
+          base = this._makeRollup(base, topk, 'OTHER');
+          dimVals = new Set(['OTHER', ...topk])
+        }
+
+        values[n] = {};
+        [...dimVals].forEach(v => {
+          values[n][v] = {
+            current: curr[v] || 0,
+            baseline: base[v] || 0
+          }
+        });
+      });
+
+      return values;
+    }
+  ),
+
+  scores: Ember.computed(
+    '_dataRollup',
+    'mode',
+    function () {
+      const { _dataRollup: data, mode } = this.getProperties('_dataRollup', 'mode');
+
+      const transformation = this._makeTransformation(mode);
+
+      const scores = {};
+      Object.keys(data).forEach(n => {
+        scores[n] = {};
+        Object.keys(data[n]).forEach(v => {
+          const curr = data[n][v].current;
+          const base = data[n][v].baseline;
+          const currTotal = this._makeSum(data[n], (d) => d.current);
+          const baseTotal = this._makeSum(data[n], (d) => d.baseline);
+
+          scores[n][v] = Math.round(transformation(curr, base, currTotal, baseTotal) * 10000) / 100.0; // percent, 2 commas
+        })
+      });
+
+      return scores;
+    }
+  ),
+
+  contributions: Ember.computed(
+    '_dataRollup',
+    function () {
+      const { _dataRollup: data } = this.getProperties('_dataRollup');
+
+      const contributions = {};
+      Object.keys(data).forEach(n => {
+        scores[n] = {};
+        Object.keys(data[n]).forEach(v => {
+          const curr = data[n][v].current;
+          const currTotal = this._makeSum(data[n], (d) => d.current);
+
+          sizes[n][v] = Math.round(1.0 * curr / currTotal * 10000) / 100.0; // percent, 2 commas
+        })
+      });
+
+      return contributions;
+    }
+  ),
+
+  _makeTransformation(mode) {
+    switch (mode) {
+      case 'change':
+        return (curr, base, currTotal, baseTotal) => curr / base - 1;
+      case 'contributionDiff':
+        return (curr, base, currTotal, baseTotal) => curr / currTotal - base / baseTotal;
+      case 'contributionToDiff':
+        return (curr, base, currTotal, baseTotal) => (curr - base) / baseTotal;
+    }
+    return (curr, base, currTotal, baseTotal) => 0;
+  },
+
+  _makeRollup(dimNameObj, topk, otherValue) {
+    if (!dimNameObj) {
+      return dimNameObj;
+    }
+    const rollup = {};
+    [...topk].forEach(v => rollup[v] = dimNameObj[v]);
+
+    const sumOther = this._makeSumOther(dimNameObj, topk);
+    rollup[otherValue] = sumOther;
+
+    return rollup;
+  },
+
+  _makeSum(dimNameObj, funcExtract) {
+    if (!dimNameObj) {
+      return 0;
+    }
+    return Object.values(dimNameObj).reduce((agg, x) => agg + funcExtract(x), 0);
+  },
+
+  _makeSumOther(dimNameObj, topk) {
```

Comment:
What are `dimNamObj` and `topk`? Is there a way we can rename these variables to make them more meaningful?

Reply:
hmm not really without blowing this helper out of proportion. "topk" is the set of topk dimension names. "dimNameObj" only makes sense in connection with the code that calls this helper.